vmngclient/api/software_action_api.py

```diff
@@ -0,0 +1,186 @@
+
+
+class Family(Enum):
+    CEDGE = "vedge"
+    VBOND = "vedge"
+    VEDGE = "vedge"
+    VSMART = "vedge"
+    VMANAGE = "vmanage"
+
+
+class VersionType(Enum):
+    VSMART = "vmanage"
+    VBOND = "vmanage"
+    VEDGE = "vmanage"
+    CEDGE = "vmanage"
+    VMANAGE = "vmanage"
+
+
+class DeviceType(Enum):
+    VSMART = "controller"
+    VBOND = "controller"
+    VEDGE = "vedge"
+    CEDGE = "vedge"
+    VMANAGE = "vmanage"
+
+
+@define
+class InstallSpecification:
+
+    family: Family
+    version_type: VersionType
+    device_type: DeviceType
+
+
+class SoftwareActionAPI:
+    """
+    API methods for software actions. All methods
+    are exececutable on all device categories.
+    """
+
+    def __init__(self, repository: RepositoryAPI) -> None:
+
+        self.repository = repository
+
+    def activate_software(self, version_to_activate: str) -> str:
+        """
+        Method to set choosen version as current version
+
+        Args:
+            version_to_activate (str): version to be set as current version
+
+        Returns:
+            str: action id
+        """
+        self.repository.complete_device_list(version_to_activate, "available_versions")
+        url = "/dataservice/device/action/changepartition"
+        payload = {
+            "action": "changepartition",
+            "devices": self.repository.devices,
+            "deviceType": "vmanage",
+        }
+        activate = dict(self.repository.session.post_json(url, payload))
+        return activate["id"]
+
+    def upgrade_software(
+        self,
+        software_image: str,
+        install_spec: InstallSpecification,
+        reboot: bool,
+        sync: bool = True,
+    ) -> str:
+        """
+        Method to install new software
+
+        Args:
+            software_image (str): path to software image
+            install_spec (InstallSpecification): specification of devices
+            on which the action is to be performed
+            reboot (bool): reboot device after action end
+            sync (bool, optional): Synchronize settings. Defaults to True.
+
+        Raises:
+            ValueError: Raise error if downgrade in certain cases
+
+        Returns:
+            str: action id
+        """
+        self.install_spec = install_spec
+
+        url = "/dataservice/device/action/install"
+        payload: Dict[str, Any] = {
+            "action": "install",
+            "input": {
+                "vEdgeVPN": 0,
+                "vSmartVPN": 0,
+                "family": self.install_spec.family,
+                "version": self.repository.get_image_version(software_image),
+                "versionType": self.install_spec.version_type,
+                "reboot": reboot,
+                "sync": sync,
+            },
+            "devices": self.repository.devices,
+            "deviceType": self.install_spec.device_type,
+        }
+
+        if self.install_spec.family in (Family.VMANAGE.value, Family.CEDGE.value):
```

Comment:
Maybe create some special types?
```python
class Family(Flag):
...    SPECIAL = VMANAGE  | CEDGE
```

```suggestion
        if self.install_spec.family in Family.SPECIAL:
```

Reply:
I had to create another dataclass to differentiate vmanage and cedge devices. Is it good idea to create another Enum class? I mean, there is quite big amount of them in this module right now.